runtime/autoload/provider/clipboard.vim

```diff
@@ -5,42 +5,81 @@ let s:copy = {}
 let s:paste = {}
 
 function! s:try_cmd(cmd, ...)
-  let out = a:0 ? systemlist(a:cmd, a:1, 1) : systemlist(a:cmd, [''], 1)
+  let argv = split(a:cmd, " ")
+  let out = a:0 ? systemlist(argv, a:1, 1) : systemlist(argv, [''], 1)
   if v:shell_error
-    echo "clipboard: error: ".(len(out) ? out[0] : '')
-    return ''
+    echohl WarningMsg "clipboard: error: ".(len(out) ? out[0] : '')
+    return 0
   endif
   return out
 endfunction
 
+let s:cache_enabled = 1
 if executable('pbcopy')
   let s:copy['+'] = 'pbcopy'
   let s:paste['+'] = 'pbpaste'
   let s:copy['*'] = s:copy['+']
   let s:paste['*'] = s:paste['+']
+  let s:cache_enabled = 0
 elseif executable('xclip')
-  let s:copy['+'] = 'xclip -i -selection clipboard'
+  let s:copy['+'] = 'xclip -quiet -i -selection clipboard'
   let s:paste['+'] = 'xclip -o -selection clipboard'
-  let s:copy['*'] = 'xclip -i -selection primary'
+  let s:copy['*'] = 'xclip -quiet -i -selection primary'
   let s:paste['*'] = 'xclip -o -selection primary'
 elseif executable('xsel')
-  let s:copy['+'] = 'xsel -i -b'
+  let s:copy['+'] = 'xsel --nodetach -i -b'
   let s:paste['+'] = 'xsel -o -b'
-  let s:copy['*'] = 'xsel -i -p'
+  let s:copy['*'] = 'xsel --nodetach -i -p'
   let s:paste['*'] = 'xsel -o -p'
 else
   echom 'clipboard: No shell command for communicating with the clipboard found.'
+  echom 'clipboard: see :help nvim-clipboard'
   finish
 endif
 
+let s:selection = { 'owning': 0, 'data': [] }
+
+function! s:selection.on_exit(jobid, data, event)
+  " At this point this nvim instance might already have launched
+  " a new provider instance. Don't drop ownership in this case.
+  if self.owning == a:jobid
+    let self.owning = 0
+  endif
+endfunction
+
+let s:selections = { '*': s:selection, '+': copy(s:selection)}
+
 let s:clipboard = {}
 
 function! s:clipboard.get(reg)
+  if s:selections[a:reg].owning > 0
+    return s:selections[a:reg].data
+  end
   return s:try_cmd(s:paste[a:reg])
 endfunction
 
 function! s:clipboard.set(lines, regtype, reg)
-  call s:try_cmd(s:copy[a:reg], a:lines)
+  if s:cache_enabled == 0
+    call s:try_cmd(s:copy[a:reg], a:lines)
+    return
+  end
+
+  let selection = s:selections[a:reg]
+  if selection.owning > 0
+    " The previous provider instance should exit when the new one takes
+    " ownership, but kill it to be sure we don't fill up the job table.
+    call jobstop(selection.owning)
+  end
+  let selection.data = [a:lines, a:regtype]
+  let argv = split(s:copy[a:reg], " ")
+  let jobid = jobstart(argv, selection)
+  if jobid <= 0
+    echohl WarningMsg "clipboard: error when invoking provider"
```

Comment:
Hmm, in the second case nvim by itself `E902: "xclip" is not an executable`. So I would just leave it like it is.